oauth2_http/java/com/google/auth/oauth2/ImpersonatedCredentials.java

```diff
@@ -0,0 +1,405 @@
+      throw new IllegalStateException(LIFETIME_EXCEEDED_ERROR);
+    }
+  }
+
+  /**
+   * @param sourceCredentials = Source Credentials.
+   * @param targetPrincipal   = targetPrincipal;
+   * @param scopes            = scopes;
+   * @param lifetime          = lifetime;
+   * @param transportFactory  = HTTP transport factory, creates the transport used
+   *                          to get access tokens.
+   * @deprecated Use {@link #create(ImpersonatedCredentials)} instead. This constructor
+   *             will either be deleted or made private in a later version.
+   */  
+  @Deprecated
+  ImpersonatedCredentials(GoogleCredentials sourceCredentials, String targetPrincipal, List<String> scopes,
+      int lifetime, HttpTransportFactory transportFactory) {
+    this(sourceCredentials, targetPrincipal, new ArrayList<String>(), scopes, lifetime, transportFactory);
+  }
+
+  /**
+   * @param sourceCredentials = Source Credentials.
+   * @param targetPrincipal   = targetPrincipal;
+   * @param delegates         = delegates;
+   * @param scopes            = scopes;
+   * @param lifetime          = lifetime;
+   * @deprecated Use {@link #create(ImpersonatedCredentials)} instead. This constructor
+   *             will either be deleted or made private in a later version.
+   */  
+  @Deprecated
+  ImpersonatedCredentials(GoogleCredentials sourceCredentials, String targetPrincipal, List<String> scopes,
+      int lifetime) {
+    this(sourceCredentials, targetPrincipal, new ArrayList<String>(), scopes, lifetime, null);
+  }
+
+  @Override
+  public AccessToken refreshAccessToken() throws IOException {
+    if (this.sourceCredentials.getAccessToken() == null) {
+      this.sourceCredentials = this.sourceCredentials.createScoped(Arrays.asList(CLOUD_PLATFORM_SCOPE));
+      this.sourceCredentials.refresh();
+    }
+    if (this.sourceCredentials.getAccessToken().getExpirationTime().before(new Date())) {
+      this.sourceCredentials.refresh();
+    }
+
+    HttpTransport httpTransport = this.transportFactory.create();
+    JsonObjectParser parser = new JsonObjectParser(OAuth2Utils.JSON_FACTORY);
+
+    HttpCredentialsAdapter adapter = new HttpCredentialsAdapter(sourceCredentials);
+    HttpRequestFactory requestFactory = httpTransport.createRequestFactory();
+
+    String endpointUrl = String.format(IAM_ENDPOINT, this.targetPrincipal);
+    GenericUrl url = new GenericUrl(endpointUrl);
+
+    Map<String, Object> body = ImmutableMap.<String, Object>of("delegates", this.delegates, "scope", this.scopes,
+        "lifetime", this.lifetime + "s");
+
+    HttpContent requestContent = new JsonHttpContent(parser.getJsonFactory(), body);
+    HttpRequest request = requestFactory.buildPostRequest(url, requestContent);
+    adapter.initialize(request);
+    request.setParser(parser);
+
+    HttpResponse response = request.execute();
+    int statusCode = response.getStatusCode();
+
+    if (statusCode == HttpStatusCodes.STATUS_CODE_UNAUTHORIZED) {
```

Comment:
We should use the http client's error handling capabilities, but that can be done later.

Reply:
andled all errors as `IOException` and added prefix to indicate where its from